src/main/java/ch/njol/skript/expressions/ExprXYZComponent.java

```diff
@@ -0,0 +1,196 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import org.bukkit.event.Event;
+import org.bukkit.util.Vector;
+import org.eclipse.jdt.annotation.Nullable;
+import org.joml.Quaternionf;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.classes.Changer.ChangerUtils;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+
+/**
+ * Ported by Sashie from skript-vectors with bi0qaw's permission.
+ * @author bi0qaw
+ */
+@Name("Vector/Quaternion - XYZ Component")
+@Description("Gets or changes the x, y or z component of <a href='classes.html#vector'>vectors</a>/<a href='classes.html#quaternion'>quaternions</a>.")
+@Examples({
+	"set {_v} to vector 1, 2, 3",
+	"send \"%x of {_v}%, %y of {_v}%, %z of {_v}%\"",
+	"add 1 to x of {_v}",
+	"add 2 to y of {_v}",
+	"add 3 to z of {_v}",
+	"send \"%x of {_v}%, %y of {_v}%, %z of {_v}%\"",
+	"set x component of {_v} to 1",
+	"set y component of {_v} to 2",
+	"set z component of {_v} to 3",
+	"send \"%x component of {_v}%, %y component of {_v}%, %z component of {_v}%\""
+})
+@Since("2.2-dev28, INSERT VERSION (Quaternions)")
+public class ExprXYZComponent extends SimplePropertyExpression<Object, Number> {
+
+	static {
+		String types = "vectors";
+		if (Skript.isRunningMinecraft(1, 19, 4))
+			types += "/quaternions";
+		register(ExprXYZComponent.class, Number.class, "[vector|quaternion] (0¦x|1¦y|2¦z) [component[s]]", types);
```

Comment:
The pattern supports x y and z as 0 1 and 2, respectively
There's no support for w, and if w is supposed to be 0, as convert() indicates, then for quaternions x = w, y = x, z = y, and z isn't accessible.